include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+   * scalar FiniteElement @projection_fe, @p mass_quadrature and @p data_quadrature
+   * Quadrature rules.
+   * First, the data located at @p data_quadrature of each cell is L2-projected
+   * to the continuous space defined by a single FiniteElement @projection_fe .
+   * In  doing so the mass matrix of this element is required, which will be calculated
+   * with the @p mass_quadrature rule . Should the cell now belong to another processor,
+   * the data is then sent to this processor. The class makes use of a feature of p4est
+   * (and parallel::distributed::Triangulation) that allows one to attach information
+   * to cells during mesh refinement and rebalancing.
+   * On receiving information on the target cell, the data is projected back to the
+   * quadrature points. In the case that local refinement is performed, we have to first
+   * project local DoF values of the parent element to each child.
+   *
+   * This class is templated by @p DataType type, however the user's @p DataType class
+   * has to be derived from the TransferableQuadraturePointData class. In practice
+   * that amounts to implementing the following three functions shown below for a
+   * quadrature point data with 2 scalars :
+   * <code>
+   * class MyQData : public TransferableQuadraturePointData
+   * {
+   * public:
+   *   unsigned int number_of_values() const
+   *   {
+   *      return 2;
+   *   }
+   *
+   *   // a function to pack scalars into a vector
+   *   void pack_values(std::vector<double> &values) const
+   *   {
+   *     Assert (scalars.size()==2, ExcInternalError());
+   *     scalars[0] = value1;
+   *     scalars[1] = value2;
+   *   }
+   *
+   *   void unpack_values(const std::vector<double> &values)
+   *   {
+   *     Assert (scalars.size() ==2, ExcInternalError());
+   *     value1 = scalars[0];
+   *     value2 = scalars[1];
+   *   }
+   * };
+   * </code>
+   * Note that the order of packing and unpacking has to be the same.
+   *
+   * This class can then be use with CellDataStorage in the following way:
+   * <code>
+   * CellDataStorage<typename Triangulation<dim,dim>::cell_iterator,MyQData> data_storage;
+   * distributed::ContinuousQuadratureDataTransfer<dim,MyQData> data_transfer(FE_Q<dim>(2),QGauss<dim>(3),QGauss<dim>(4));
+   * //...populate data for all active cells in data_storage
+   * //...mark cells for refinement...
+   * data_transfer.prepare_for_coarsening_and_refinement(triangulation,&data_storage);
+   * triangulation.execute_coarsening_and_refinement();
+   * //...initialize quadrature point data on new cells by calling CellDataStorage::reinit()
+   * data_transfer.interpolate();
+   * </code>
+   * This approach can be extended to quadrature point data with Tensors of
+   * arbitrary order, although with a little bit more work in packing and
+   * unpacking of data inside MyQData class.
+   *
+   * @note Currently coarsening is not supported.
```

Comment:
I don't actually recall how it's done, but it uses the restriction matrices of the finite element.

You made the decision that cells can have different types (and different numbers) of data attached to them. Let's assume for a moment that that was not the case, and that every cell has the same number and type of data attached. Then one way to facilitate the transfer in an easy way would have been to create a `DoFHandler` object, project the data from the quadrature points to the FE space, use `SolutionTransfer` to get the data to the correct processor and interpolated to the correct cell, and then just evaluate the fields at the quadrature points again. That would have been pretty easy to implement, but you don't have that option with your design.

But I think you can look at the design in `SolutionTransfer` to see how it's done there. My suspicion is that we first interpolate to the parent cell (after `prepare_coarsening_and_refinement()`), and only then do the transfer to a different processor. I may be wrong, but that's the line I would investigate.